xlators/mgmt/glusterd/src/glusterd-op-sm.c

```diff
@@ -2821,6 +2823,136 @@ glusterd_set_shared_storage(dict_t *dict, char *key, char *value,
+               "Invalid value entry");
+        goto out;
+    }
+
+    if (bool_val_1 == bool_val_2)
+        ret = 1;
+    else
+        ret = 0;
+
+out:
+    return ret;
+}
+
+int
+glusterd_compare_option_value(glusterd_volinfo_t *volinfo, char *input_key,
+                              char *input_value)
+{
+    struct volopt_map_entry *vme = NULL;
+    int ret = -1;
+    char *def_val = NULL;
+    int keylen;
+    glusterd_conf_t *priv = NULL;
+    xlator_t *this = THIS;
+
+    priv = this->private;
+    GF_VALIDATE_OR_GOTO(this->name, priv, out);
+
+    for (vme = &glusterd_volopt_map[0]; vme->key; vme++) {
+        if (!strcmp(vme->key, input_key)) {
+            /* First look for the key in the priv->opts for global option
+             * and then into vol_dict, if its not present then look for
+             * translator default value */
+            keylen = strlen(vme->key);
+            ret = dict_get_strn(priv->opts, vme->key, keylen, &def_val);
+            if (!def_val) {
+                ret = dict_get_strn(volinfo->dict, vme->key, keylen, &def_val);
+                if (ret == -ENOENT)
+                    def_val = glusterd_get_option_value(volinfo, vme->key);
+                if (!def_val) {
+                    if (vme->value) {
+                        def_val = vme->value;
+                    } else {
+                        ret = glusterd_get_value_for_vme_entry(vme, &def_val);
+                        if (ret)
+                            goto out;
+                        else if (ret == -2)
+                            continue;
+                    }
+                }
+            }
+        }
+    }
+
+    ret = glusterd_compare_bool_value(input_value, def_val);
+    def_val = NULL;
+
+out:
+    if (def_val)
+        GF_FREE(def_val);
+
+    return ret;
+}
+
+int
+glusterd_dependency_chain_check(glusterd_volinfo_t *volinfo, char *key,
+                                char *value, char **errstr)
+{
+    int ind = 0;
+    int j = 0;
+    int match_cnt = 0;
+    int ret = -1;
+    char *err_s = NULL;
+    while (prohibited_clauses[ind]) {
+        if (!prohibited_clauses[ind][0].op)
+            break;
+        if (strcmp(prohibited_clauses[ind][0].val, CLAUSE_LABEL) == 0) {
+            err_s = gf_strdup(prohibited_clauses[ind][0].op);
+        } else {
+            match_cnt = 0;
+            j = 0;
+            while (j >= 0) {
+                if (!prohibited_clauses[ind][j].op)
+                    break;
+                if (key && strcmp(key, prohibited_clauses[ind][j].op) == 0) {
+                    ret = glusterd_compare_bool_value(
+                        value, prohibited_clauses[ind][j].val);
+                } else {
+                    ret = glusterd_compare_option_value(
+                        volinfo, prohibited_clauses[ind][j].op,
+                        prohibited_clauses[ind][j].val);
+                    if (ret == 1)
+                        match_cnt++;
```

Comment:
Also it should be handled as disctinguished error if the runtime value for the `prohibited_clauses[ind][j].op` (whether in volinfo or in dict) is non-Boolean. If this occurs,
- evaluation should be stopped and conclude on error right away
- use a specific error message indicating type mismatch.

Reply:
If an option is supposed to take it in a boolean value only, then the volinfo will never have the incorrect value. It is checked before adding to it.

Also, in a runtime `vol set` operation, the sanity check of the `option` and `value` passed is done before reaching here, so we won't face the issue.

And, I think the type check of the value for the `option` in the predicate can be ignored, as it will be added by a developer while adding the dependency so, it will be taken care of by him?